display/ssd1306.h

```diff
@@ -312,6 +328,8 @@ class SSD1306 : public I2CDevice, Looper, StateMachine, SaberBase, private Audio
 
   void SB_On() override {
     ShowFile(&IMG_on, font_config.ProffieOSOnImageDuration);
+    // Make idle.bmp wait until first time blade on
```

Comment:
but why?

Reply:
This is echoed in the comment above where we need non-SD card content initially so that Mass Storage option will work. Otherwise, the board never releases the SD card for mounting to the desktop. Plus, I wasn't looking to completely hide the battery bar. It's nice to have it at boot to see where the battery level is at.

Reply:
I didn't get that part from the comments (which means that the comments needs to be improved.)
Also, I think we might need the SD card to be released if you plug it into USB after turning it on and off too.
Maybe we just play idle immediately on boot but stop doing it as soon as USB is connected IF mass storage is on...

Reply:
Also, the whole naming convention of "idle.bmp" is somewhat conflicting with the OffType OFF_IDLE, no? I realized that once I got down to the screen sleeping part.

Reply:
I think idle.bmp is fine.
The other option is just "off.bmp" I guess.

Reply:
> I didn't get that part from the comments (which means that the comments needs to be improved.)
Updated - added "so Mass Storage can work" there as well.

> Also, I think we might need the SD card to be released if you plug it into USB after turning it on and off too.
Just typing reset in serial monitor or toggling a kill key/switch remounts the SD card. Do you feel that is too much to ask end users?
> Maybe we just play idle immediately on boot but stop doing it as soon as USB is connected IF mass storage is on...
I really think the battery meter is still nice to have at boot, no? But stopping idle.bmp playing on USB connection is cool, yes. I'm just not familiar enough with the code for detecting USB communication. Something like:
if (USBD_Connected() BECOMES_TRUE && MASS STORAGE IS ON... MTP SOMETHING?**) then stop the loop.
I feel like it's a double negative and I'm on the wrong course.

Reply:
> Just typing reset in serial monitor or toggling a kill key/switch remounts the SD card. Do you feel that is too much to ask end users?

Yes, it is inconvenient. Also, the reset command tends to cause problems if used incorrectly, so I don't want to rely on everybody using that.

I don't idle.bmp should start after boot. It could be nice to have an option to show the battery meter first, but some people might not want that if it doesn't go with the design they have in mind....

For the condition, I think all we need is:

#ifdef USB_CLASS_MSC
   if (USBD_Connected()) { /* stop loop here */ }
#endif

I don't think we need to worry about MTP since MTP works without unmounting the SD card.

Reply:
I think there's another piece to the puzzle I don't have. The piece you provided above looks simple enough, but I think it needs expansion for it to work on both connect and disconnect USB. The closest I can get to having USB trigger the battery meter on/off instead of idle.bmp works like: With no Mass Storage option, all is well. idle.bmp all when expected. Turning Mass Storage on, when boot/upload with USB connected, it shows batt, SD card mounts to computer fine, pulling usb does idle again. But having it stop idle.bmp when reconnecting USB is eluding me. I have USB detection in the loop for SCREEN_PLI section, (the seems to be working part), but something similar must have to go in the SCREEN_IMAGE loop, as idle.bmp would need to get interrupted there when USB is connected. I have tried the conditions in both area, variants of it, etc... and inexplicably I see the battery meter when SB_Off happens even though the conditions are not met, and/or I see a blip of idle.bmp play but then it goes back to battery meter. Tough to word. Is there another condition, such as when the SD card gets mounted/ unmounted that could be used for detection? 

The commented out parts is where I've been unsuccessful and playing around with the reconnect part. The SCREEN_PLI section seems to handle unplugging correctly. And why would SB_OFF be showing battery meter at all if no USB connected?
Commiting update now.

Reply:
Ok. update. I think I found why I'm going crazy here. 
It seems that USB is not being detected reliably as connected or not.
I added some prints to the void Loop to monitor the status via a serial connection while I plugged in/out USB cable.
It is reporting connected although it is not, with a single report of truth (not connected) approx every 1 minute.
This is what's throwing off my attempts above, and I'm not sure if it's a buffer thing, or if we'd need to prioritize the USBD_Connected() polling somehow?
Here's the results of the monitoring. This is all completely with USB unplugged:

22:02:28.360 -> --------------------while(true)
22:02:28.360 -> --------------------USBD_Connected
22:02:29.337 -> --------------------while(true)
22:02:29.337 -> --------------------USBD NOT Connected
22:02:30.351 -> --------------------while(true)
22:02:30.351 -> --------------------USBD_Connected
22:02:31.368 -> --------------------while(true)
22:02:31.368 -> --------------------USBD_Connected

A summary of just the NOT connected prints shows these irregular intervals:
21:59:49.209 -> --------------------USBD NOT Connected
22:01:09.269 -> --------------------USBD NOT Connected
22:02:29.337 -> --------------------USBD NOT Connected
22:03:49.413 -> --------------------USBD NOT Connected
22:04:20.175 -> --------------------USBD NOT Connected
22:04:57.214 -> --------------------USBD NOT Connected
22:05:40.254 -> --------------------USBD NOT Connected

Thoughts?

Reply:
That is weird and unexpected.
Can I see the code you used to print that?

Reply:
Yes, it's updated in the latest commit.
Right under where Sidpay is Initialized, I found that that loop is running contantly in tandem with the Screen switch, so I stuck it in there  ¯\_(ツ)_/¯

    while (true) {
      STDOUT << "--------------------while(true)\n";
      if (USBD_Connected()) {
        STDOUT << "--------------------USBD_Connected\n";
      } else {
        STDOUT << "--------------------USBD NOT Connected\n";
      }
      millis_to_display_ = FillFrameBuffer();
      frame_start_time_ = millis();
      lock_fb_ = true;
      frame_available_ = false;

Reply:
FYI I never get NOT connected message when it's in it's own if condition like this instead:

  while (true) {
      STDOUT << "--------------------while(true)\n";
      if (USBD_Connected()) {
        STDOUT << "--------------------USBD_Connected\n";
      } 
      if (!USBD_Connected()) {
        STDOUT << "--------------------USBD NOT Connected\n";
      }
      millis_to_display_ = FillFrameBuffer();
      frame_start_time_ = millis();

Reply:
That doesn't make any sense at all.

Reply:
Agreed. Been pulling my hair out for a week on this.
When USB is unplugged, the single instance of NOT connected triggers idle.bmp, which plays once completely until it finishes its frames. But during that playback, USBD_Connected() has reverted back to true in the background for some reason, so at END OF LOOP, when we go back to SCREEN_PLI, we now get DrawBatteryBar instead until the next single instance of NOT connected...and repeat the cycle.

Reply:
How about you just make it work and show the idle.bmp all the time and I'll figure out how to make it not do it when USB is connected?

Reply:
Well sure that's super easy at this point. To clarify, this would be intentionally submitting something that breaks mass storage in lieu of you figuring out that part, as opposed to just reverting back to where I had idle.bmp just waiting until first ignition?